File: 0105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
//0ms

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode *tree(unordered_map<int,int> &m,vector<int> &p,int i,int j,int &k)
    {
        if(i>j)
            return NULL;
        int n=p[k++];
        TreeNode *root=new TreeNode(n);
        root->left=tree(m,p,i,m[n]-1,k);
        root->right=tree(m,p,m[n]+1,j,k);
        return root;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) 
    {
        unordered_map<int,int> m;
        for(int i=0;i<inorder.size();i++)
            m[inorder[i]]=i;
        int k=0;
        return tree(m,preorder,0,inorder.size()-1,k);
    }
};

File: 0101-0200/0105.Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.js
//0ms

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
  let preNdx = 0;
  let inNdx = 0;

  function dfs(limit) {
    if (preNdx >= preorder.length) return null;
    if (inorder[inNdx] === limit) {
      inNdx++;
      return null;
    }

    const root = new TreeNode(preorder[preNdx++]);

    root.left = dfs(root.val);
    root.right = dfs(limit);

    return root;
  }

  return dfs(Infinity);
};
